### PR TITLE
support ipip

### PR DESCRIPTION
**Pull request by ishidawataru**

### CODE REVIEW

File: ipam.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	"sync"
+
+	etcd "github.com/coreos/etcd/client"
+	"github.com/osrg/gobgp/table"
+	"golang.org/x/net/context"
+)
+
+type ipPool struct {
+	CIDR string `json:"cidr"`
+	IPIP string `json:"ipip"`
```
Comment (robbrockbank) [inline ipam.go:32]: An FYI - we've also now added an IPIP mode that allows use of IPIP only when the originator of the route is on a different subnet.  Doesn't need to addressed as part of this PR though.

File: ipam.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	watcher := api.Watcher(CALICO_IPAM, &etcd.WatcherOptions{Recursive: true, AfterIndex: index})
+	for {
+		res, err := watcher.Next(context.Background())
+		if err != nil {
+			return err
+		}
+		del := false
+		node := res.Node
+		switch res.Action {
+		case "set", "create":
```
Comment (robbrockbank) [inline ipam.go]: Need to handle update, compareAndSwap for completeness?

File: ipam.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	CIDR string `json:"cidr"`
+	IPIP string `json:"ipip"`
+}
+
+type ipamCache struct {
+	mu sync.RWMutex
+	m  map[string]*ipPool
+}
+
+func (c *ipamCache) add(p *ipPool, del bool) bool {
```
Comment (robbrockbank) [inline ipam.go]: Some func docs would be useful here.   For example, i"m not sure what the returning bool represents.

Comment (robbrockbank) [reply ipam.go]: Seems odd to have an add function with a del boolean.

Would it be clearer just to have separate add and del functions, or perhaps rename to update()

File: ipam.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		} else {
+			c.m[p.CIDR] = p
+			return true
+		}
+	}
+	c.m[p.CIDR] = p
+	return false
+}
+
+func (c *ipamCache) get(prefix string) *ipPool {
```
Comment (robbrockbank) [inline ipam.go]: Is this function used?  If so, does it need to handle the prefix not being in the map?

File: ipam.go (hunk 3 above)
Comment (robbrockbank) [inline ipam.go]: Hmmm, actually from reading the code below - should this actually be called update() rather than add() ?

**Comment by robbrockbank** [comment]
Some merge conflicts now on this PR, and a couple of outstanding review comments.

**Comment by ishidawataru** [comment]
@robbrockbank Thanks for your review. Reflected your comments.

I also added the support for IPIP mode.  PTAL.

**Comment by robbrockbank** [comment]
I'm happy to get this in - but looks like there is a merge conflict - probably because of the earlier merge.

**Comment by ishidawataru** [comment]
fixed the conflict

**Comment by robbrockbank** [comment]
lgtm.

**Pull request merged**